### PR TITLE
Separate program values from their representations

### PR DESCRIPTION
**Pull request by s-kybound**
This PR extends the Result type with an optional Representation field, as long as a Result Value is present. This is done to allow alternate languages (aside from JS) to provide their own data representations for values whilst preserving the actual data passed. 

For example, arrays in JavaScript are represented as `[1, 2, 3, 4]`. The corresponding vector (array equivalent in scheme) should be represented as `#(1 2 3 4)`. This PR allows the new representation to be stored separately from the value returned (which is still a JS array).

After a program is run, the result will be mapped by the mapResult() function, which checks the language used in the context to run the appropriate result mapper. Ideally, languages will add their own logic into the mapResult() function (located in src/alt-langs/mapper.ts).

This PR is also non-invasive. If an alternate representation is not needed (the case for Source/JS/TS) or not implemented, js-slang defaults to using the original representation for the value.

(as a side note, this PR also tags all scheme pairs so that they are distinguishable from scheme vectors (despite both being represented using arrays) - a possible enhancement that can be brought to js-slang too?)

### CODE REVIEW

File: src/repl/repl.ts [hunk 1]
```diff
@@ -35,7 +36,9 @@ function startRepl(
           eval: (cmd, unusedContext, unusedFilename, callback) => {
             runInContext(cmd, context, options).then(obj => {
               if (obj.status === 'finished' || obj.status === 'suspended-non-det') {
-                callback(null, obj.value)
+                // if the result is a representation, display the representation.
+                // else, default to standard value representation.
+                callback(null, obj.representation ? obj.representation : obj.value)
```
Comment (RichDom2185) [inline src/repl/repl.ts:41]: Should we check for `!==` undefined here instead of truthy/falsy?

File: src/repl/repl.ts [hunk 2]
```diff
@@ -97,7 +102,7 @@ function main() {
         'default'
       ],
       ['h', 'help', 'display this help'],
-      ['e', 'eval', "don't show REPL, only display output of evaluation"]
+      ['e', 'eval', "don't show REPL, only dispResultlay output of evaluation"]
```
Comment (RichDom2185) [inline src/repl/repl.ts:105]: Typo

File: src/types.ts [hunk 3]
```diff
@@ -275,6 +276,10 @@ export interface Finished {
   status: 'finished'
   context: Context
   value: Value
+  representation?: Representation // if the returned value needs a unique representation,
+  // (for example if the language used is not JS),
+  // the display of the result will use the representation
+  // field instead
```
Comment (RichDom2185) [inline src/types.ts:282]: I'm thinking if we can even use this to handle #1124

File: src/repl/repl.ts (hunk 2 above)
Comment (s-kybound) [reply src/repl/repl.ts:105]: Thanks!

File: src/repl/repl.ts (hunk 1 above)
Comment (s-kybound) [reply src/repl/repl.ts:41]: Will do!

File: src/types.ts (hunk 3 above)
Comment (s-kybound) [reply src/types.ts:282]: Sorry, could you elaborate? 😅 Representations currently only concern the final return value of a program - for now, for functions like display() or display_list(), the output remains unaffected

**Review by RichDom2185 — APPROVED** [review]
LGTM, thanks!

**Review by martin-henz — APPROVED** [review]
Neat integration of Scheme into js-slang.

**Pull request merged**